### PR TITLE
:sparkles: Strict mode

### PR DESCRIPTION
**Pull request by antoinecarat**
This is adding strict-mode at every level of our mechanism. Allowing to choose if you want to fail safely (default behaviour) or to throw an error as soon as something unexpected occurs.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,6 +107,7 @@ type Formatter = (value: any, ...params: any[]) => string
 - `name` is the name of the newly created and interpolated google doc file.
 - `data` is the data given for the [interpolation](#interpolation)
 - `formatters` will be used for [interpolation](#interpolation)
+- `strict` indicates whether to use [sctrict mode](#strict-mode) or not.
```
Comment (Errorname) [inline README.md]: `sctrict`

File: README.md [hunk 2]
```diff
@@ -116,12 +118,14 @@ interface DiscoveryOptions {
...
   data?: Object
   formatters?: Formatters
+  strict?: boolean
 }
 ```
 
 - `source` is the ID of the file which will be interpolated.
 - `data` is the data given for the [interpolation](#interpolation)
 - `formatters` will be used for [interpolation](#interpolation)
+- `strict` indicates whether to use [sctrict mode](#strict-mode) or not.
```
Comment (Errorname) [inline README.md]: `sctrict`

File: README.md [hunk 3]
```diff
@@ -244,7 +248,13 @@ Hi COURTOISON. Today is Tuesday, tomorrow is Wednesday.
 - **image(width, height)**: Transform an url string to an image
 - _More to come..._
 
-_Warning: If you use an undefined formatter it will be simply ignored, which could lead to unexpected results if you're chaining them._
+_Warning: If you use an undefined formatter it will be simply ignored, which could lead to unexpected results if you're chaining them. See [Strict mode](#strict-mode)_
+
+### Strict mode
+
+By default, _google-docs-mustaches_ is failing safely, which means that you don't have to worry about using an undefined variable or a unknow formatter, the generated errors will be catched and treated by the program itself. However, you might face unexepected behaviour if for example, you chain several formatters and one of them is misspelled, it would be ignored and the output of your formatters pipeline won't match your expectations.
```
Comment (Errorname) [inline README.md]: "`an` unkown formatter"
`unexepected`

File: README.md [hunk 4]
```diff
@@ -244,7 +248,13 @@ Hi COURTOISON. Today is Tuesday, tomorrow is Wednesday.
...
+
+To avoid this, we provide a strict mode for `.interpolate` and `.discovery`. Instead of using an empty string or your fallback when encountering an undefined variable, it will throw an exception, aborting immediatly the interpolation of your document.
```
Comment (Errorname) [inline README.md]: `immediatly`

Comment (Errorname) [inline README.md]: Should we rename `discovery` to `discover`?

File: src/Mustaches.test.js [hunk 5]
```diff
@@ -161,6 +192,35 @@ describe('mustaches', () => {
     expect(placeholders).toMatchSnapshot()
   })
 
+  test('discovery with strict mode: Undefined variable', async () => {
```
Comment (Errorname) [inline src/Mustaches.test.js]: For consistency, I would name the test: "discovery with strict mode - undefined variable"

File: src/interpolation/compute/compute.ts [hunk 6]
```diff
@@ -2,10 +2,12 @@ import dot from './dot'
 import pipe from './pipe'
 import { Placeholder, ContentPlaceholder, Formatters } from '../types'
 
+const DEFAULT_FALLBACK = ''
```
Comment (Errorname) [inline src/interpolation/compute/compute.ts]: 👍

File: src/interpolation/interpolate.test.js [hunk 7]
```diff
@@ -53,6 +54,189 @@ describe('interpolate', () => {
     expect(result).toMatchSnapshot()
     expect(updates).toMatchSnapshot()
   })
+
+  test('full document — strict: Unknown formatter', () => {
```
Comment (Errorname) [inline src/interpolation/interpolate.test.js]: Same consistency comment: "full document - strict - unknown formatter"

File: src/interpolation/interpolate.test.js [hunk 8]
```diff
@@ -53,6 +54,189 @@ describe('interpolate', () => {
...
+    const document = {
```
Comment (Errorname) [inline src/interpolation/interpolate.test.js]: Can't we reuse the same document instead of rewriting it on every test?

Comment (antoinecarat) [reply src/interpolation/interpolate.test.js]: Yes, we can.

File: README.md (hunk 4 above)
Comment (antoinecarat) [reply README.md]: I'd do it in a "pre-release refactoring", it's not really the point of this PR, right?

Comment (Errorname) [reply README.md]: I'm fine with that

File: src/interpolation/interpolate.test.js (hunk 8 above)
Comment (antoinecarat) [reply src/interpolation/interpolate.test.js]: Actually we can't as they are not the same

Comment (Errorname) [reply src/interpolation/interpolate.test.js]: Alright, we can keep it then

**Pull request merged**